### PR TITLE
rework newlib directory prioritization to fix include processing for non-Zephyr toolchains

### LINKED ISSUE #17564: Missing `stdlib.h` include when C++ standard library is used.
**Describe the bug**

`stdlib.h` is not in the GCC search path when the C++ standard library is used.

**To Reproduce**

prj.conf:

```
CONFIG_NEWLIB_LIBC=y
CONFIG_CPLUSPLUS=y
```

main.cpp:

```cpp
#include <zephyr.h>
#include <misc/printk.h>
#include <array>


void main(void)
{
	std::array<int, 3> a{ 1, 2, 3 };

	for (const auto x: a)
	{
		printk("x: %d", x);
	}
}
```

1. mkdir build && cd build
2. cmake -GNinja -DBOARD=arduino_zero ..
3. ninja
4. Build fails because of missing include, see console output below

**Expected behavior**

The example code above should compile.

**Impact**

I can not use the C++ standard library for my current project.

**Screenshots or console output**

```cmd
> ninja
[15/101] Building CXX object CMakeFiles/app.dir/src/main.cpp.obj
FAILED: CMakeFiles/app.dir/src/main.cpp.obj 
C:\Users\username\tools\gcc-arm-none-eabi-7-2018-q2-update-win32\bin\arm-none-eabi-g++.exe  -DBUILD_VERSION=zephyr-v1.14.0 -DKERNEL -D_FORTIFY_SOURCE=2 -D__LINUX_ERRNO_EXTENSIONS__ -D__ZEPHYR__=1 -IC:/Users/username/zephyrproject/zephyr/kernel/include -IC:/Users/username/zephyrproject/zephyr/arch/arm/include -IC:/Users/username/zephyrproject/zephyr/include -IC:/Users/username/zephyrproject/zephyr/include/drivers -Izephyr/include/generated -IC:/Users/username/zephyrproject/zephyr/soc/arm/atmel_sam0/samd21 -IC:/Users/username/zephyrproject/zephyr/ext/hal/atmel/asf/sam0/include/samd21 -IC:/Users/username/zephyrproject/zephyr/ext/hal/cmsis/Include -isystem C:/Users/username/tools/gcc-arm-none-eabi-7-2018-q2-update-win32/arm-none-eabi/include -std=c++11 -Os -fcheck-new -Wno-register -fno-exceptions -fno-rtti -g -Wall -Wformat -Wformat-security -Wno-format-zero-length -imacros C:/Users/username/tmp/hello_world/build/zephyr/include/generated/autoconf.h -ffreestanding -Wno-main -fno-common -mthumb -mcpu=cortex-m0plus -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -Wno-unused-but-set-variable -fno-reorder-functions -fno-defer-pop -Werror=implicit-int -Wpointer-arith -ffunction-sections -fdata-sections -mabi=aapcs -march=armv6s-m -MD -MT CMakeFiles/app.dir/src/main.cpp.obj -MF CMakeFiles\app.dir\src\main.cpp.obj.d -o CMakeFiles/app.dir/src/main.cpp.obj -c ../src/main.cpp
In file included from c:\users\username\tools\gcc-arm-none-eabi-7-2018-q2-update-win32\arm-none-eabi\include\c++\7.3.1\ext\string_conversions.h:41:0,
                 from c:\users\username\tools\gcc-arm-none-eabi-7-2018-q2-update-win32\arm-none-eabi\include\c++\7.3.1\bits\basic_string.h:6352,
                 from c:\users\username\tools\gcc-arm-none-eabi-7-2018-q2-update-win32\arm-none-eabi\include\c++\7.3.1\string:52,
                 from c:\users\username\tools\gcc-arm-none-eabi-7-2018-q2-update-win32\arm-none-eabi\include\c++\7.3.1\stdexcept:39,
                 from c:\users\username\tools\gcc-arm-none-eabi-7-2018-q2-update-win32\arm-none-eabi\include\c++\7.3.1\array:39,
                 from ../src/main.cpp:8:
c:\users\username\tools\gcc-arm-none-eabi-7-2018-q2-update-win32\arm-none-eabi\include\c++\7.3.1\cstdlib:75:15: fatal error: stdlib.h: No such file or directory
 #include_next <stdlib.h>
               ^~~~~~~~~~
compilation terminated.
[28/101] Building C object zephyr/CMakeFiles/zephyr.dir/lib/os/printk.c.obj
ninja: build stopped: subcommand failed..
```

**Environment (please complete the following information):**
* OS: Windows 10 1903
* Toolchain: GNU ARM Embedded 7-2018-q2-update
* Zephyr: v1.14.0

**Additional context**

This might be related to issue #15603. I tried the suggested workaround, but it does not help.

I tried several Zephyr versions:
* v1.14.1-rc1
* v1.14.99 (latest master, commit 1c6952c4f6de045fc5765373ff89683afd6396a4 )
* v1.14.0-rc1

v1.14.0-rc1 works but in this version the shell is not C++ compatible yet, so there are other problems to work around.

I also tried the latest GNU ARM Embedded toolchain (8-2019-q3-update), but also no success here.

If you can point my in the right direction, I will happily provide a fix myself.

### PR DESCRIPTION
**Pull request by pabigot**
This work closes #17564 and related C++ failures.

Zephyr builds with `-ffreestanding` meaning that `__STDC_HOSTED__` is false.  Consequently the toolchain standard `<stdint.h>` is a minimal header that routes to `<stdint-gcc.h>` which defines integer types without defining the marker macro `__int64_t_defined` that signals newlib `<inttypes.h>` to provide support for `PRIu64` (#14310).  In hosted builds `<stdint.h>` instead sets some flags then delegates to the host-provided (newlib) implementation of `<stdint.h>`.

In support of the desire to have support for integer types that aren't part of the freestanding system #14312 placed the newlib include headers before the system ones, specifically to make hosted features available in a non-hosted system.

This decision breaks building C++ applications that use standard C++ include files when using a non-Zephyr toolchain like `gnumarmemb`.  The C++ header `<cstdlib>` (referenced from `<array>` in #17564) wants to `#include_next <stdlib.h>`, but the include priority has placed the directory that provides it earlier in the search path, where it can no longer be found.

The workaround succeeded with the Zephyr toolchain only because the Zephyr SDK uses a cross-compiler infrastructure technique that leaves behind and references a second copy of the base include directory that is normally needed only when building the compiler.  `<stdint.h>` is then found at that second location.

This PR switches to a solution based on updating the include path to include an intercepting header that provides the required defines to satisfy the use case of #14310 without completely breaking the toolchain standard include processing.  This should work on GCC and [clang](
http://clang.llvm.org/docs/LanguageExtensions.html#has-include-next).

Note that the test case requires a patch from #18243 to support ztest with C++.

### CODE REVIEW

**Comment by pabigot** [comment]
Background information:

To understand the details you need to know what the system header search path is for a given compiler.  Get this with:

    target-g++ -xc++ -v -E - < /dev/null

With `ZEPHYR_TOOLCHAIN_VARIANT=zephyr` the standard system header search paths are (e.g.):

    /usr/local/zephyr-sdk-0.10.2/arm-zephyr-eabi/arm-zephyr-eabi/include/c++/8.3.0
    /usr/local/zephyr-sdk-0.10.2/arm-zephyr-eabi/arm-zephyr-eabi/include/c++/8.3.0/arm-zephyr-eabi
    /usr/local/zephyr-sdk-0.10.2/arm-zephyr-eabi/arm-zephyr-eabi/include/c++/8.3.0/backward
    /usr/local/zephyr-sdk-0.10.2/arm-zephyr-eabi/lib/gcc/arm-zephyr-eabi/8.3.0/include
    /usr/local/zephyr-sdk-0.10.2/arm-zephyr-eabi/lib/gcc/arm-zephyr-eabi/8.3.0/include-fixed
    /usr/local/zephyr-sdk-0.10.2/arm-zephyr-eabi/arm-zephyr-eabi/sys-include
    /usr/local/zephyr-sdk-0.10.2/arm-zephyr-eabi/arm-zephyr-eabi/include

By adding `-isystem ${LIBC_INCLUDE_DIR}` this changes to:

    /usr/local/zephyr-sdk-0.10.2/arm-zephyr-eabi/arm-zephyr-eabi/include
    /usr/local/zephyr-sdk-0.10.2/arm-zephyr-eabi/arm-zephyr-eabi/include/c++/8.3.0
    /usr/local/zephyr-sdk-0.10.2/arm-zephyr-eabi/arm-zephyr-eabi/include/c++/8.3.0/arm-zephyr-eabi
    /usr/local/zephyr-sdk-0.10.2/arm-zephyr-eabi/arm-zephyr-eabi/include/c++/8.3.0/backward
    /usr/local/zephyr-sdk-0.10.2/arm-zephyr-eabi/lib/gcc/arm-zephyr-eabi/8.3.0/include
    /usr/local/zephyr-sdk-0.10.2/arm-zephyr-eabi/lib/gcc/arm-zephyr-eabi/8.3.0/include-fixed
    /usr/local/zephyr-sdk-0.10.2/arm-zephyr-eabi/arm-zephyr-eabi/sys-include

The contents of the zephyr toolchain path ending in `/sys-include` is identical to that of `/include` except that it lacks the `c++` subdirectory.  It is an artefact of a multi-stage toolchain bootstrap that allows libc to be built using a compiler that doesn't have a real libc.

With `ZEPHYR_TOOLCHAIN_VARIANT=gnuarmemb` the standard system header search paths are (e.g.):

    /usr/local/gcc-arm-none-eabi-8-2019-q3-update/arm-none-eabi/include/c++/8.3.1
    /usr/local/gcc-arm-none-eabi-8-2019-q3-update/arm-none-eabi/include/c++/8.3.1/arm-none-eabi
    /usr/local/gcc-arm-none-eabi-8-2019-q3-update/arm-none-eabi/include/c++/8.3.1/backward
    /usr/local/gcc-arm-none-eabi-8-2019-q3-update/lib/gcc/arm-none-eabi/8.3.1/include
    /usr/local/gcc-arm-none-eabi-8-2019-q3-update/lib/gcc/arm-none-eabi/8.3.1/include-fixed
    /usr/local/gcc-arm-none-eabi-8-2019-q3-update/arm-none-eabi/include

By adding `-isystem ${LIBC_INCLUDE_DIR}` this changes to:

    /usr/local/gcc-arm-none-eabi-8-2019-q3-update/arm-none-eabi/include
    /usr/local/gcc-arm-none-eabi-8-2019-q3-update/arm-none-eabi/include/c++/8.3.1
    /usr/local/gcc-arm-none-eabi-8-2019-q3-update/arm-none-eabi/include/c++/8.3.1/arm-none-eabi
    /usr/local/gcc-arm-none-eabi-8-2019-q3-update/arm-none-eabi/include/c++/8.3.1/backward
    /usr/local/gcc-arm-none-eabi-8-2019-q3-update/lib/gcc/arm-none-eabi/8.3.1/include
    /usr/local/gcc-arm-none-eabi-8-2019-q3-update/lib/gcc/arm-none-eabi/8.3.1/include-fixed

Note that this build does not have `/sys-include` because the ARM folks use a two-stage build where the final build doesn't include it.  Consequently `#include_next` for any file in the normal include path will be found in `/sys-include` for Zephyr toolchains, but will not be found in other toolchains that don't have duplicated standard include directories.

**Comment by pabigot** [comment]
@pfalcon Please check this branch for the use case underlying #15627.  It essentially reverts a change from #14312 that you used to justify #15937, i.e. using `-isystem` for a libc header path in order to change its priority. 

That change was incorrect for newlib, but is arguably appropriate for minimal libc.

I don't have a test case for #15627 but if your solution depended on the newlib use of `-isystem` this commit may have broken it again.  If so, I think the correct path is to add support for 
 a `zephyr_early_include_directories()` macro that uses the cmake `BEFORE` tag instead of the `SYSTEM` tag to ensure that some directory is placed earlier in the search path.  Or possibly use the technique demonstrated here to avoid name conflicts.

**Comment by pfalcon** [comment]
> Please check this branch for the use case underlying #15627. It essentially reverts a change from 

@pabigot: As a quick "first thing which comes to mind", based on wording above and "Commits 3 Files changed 11", I'd say that this is too late for 2.0.

I'll try however to look in detail as time permits.

**Comment by pabigot** [comment]
@pfalcon Only one commit, and a change to two files, fixes this bug, which I thought was in scope for this stage of the release process.  The other two commits are (1) a test case to avoid regressions, and (2) additional cleanup(bug) fixes queued in #18243 necessary to do unit tests on C++ code.

However, I'm not adamant about merging this, as long as it's OK to leave C++ broken in 2.0.0 with non-Zephyr toolchains.

File: lib/libc/newlib/include/stdint.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+#include <newlib.h>
+
+#ifdef __NEWLIB__
+/* Has this header.  Older versions do it in <stdint.h>. */
+#include <sys/_stdint.h>
+#endif /* __NEWLIB__ */
+
+/* This should work on GCC and clang. */
+#if ((__GNUC__ - 0) > 0) || __has_include_next(<stdint.h>)
```
Comment (pfalcon) [inline lib/libc/newlib/include/stdint.h]: Was effect of this `__has_include_next` thingy considered on other toolchains? I personally never saw it before, and a quick google-up show a usage wrapped in `#ifdef __has_include_next`: https://www.gnu.org/software/gcc/gcc-5/changes.html

File: CODEOWNERS [hunk 2]
```diff
@@ -368,6 +368,7 @@
 /subsys/testsuite/                        @nashif
 /subsys/usb/                              @jfischer-phytec-iot @finikorg
 /tests/                                   @nashif
+/tests/application_development/libcpp/    @pabigot
```
Comment (pfalcon) [inline CODEOWNERS]: libcpp may sounds weird to some, `cpp` is C preprocessor, and C++'s standard lib is called libstdc++ (but nope, don't call it like that, Windows and stuff).

**Comment by pfalcon** [comment]
@pabigot 

> I don't have a test case for #15627 but if your solution depended on the newlib use of -isystem this commit may have broken it again.

So, testcases are includes in a PR that fixed that issue, https://github.com/zephyrproject-rtos/zephyr/pull/15937 , specifically, https://github.com/zephyrproject-rtos/zephyr/pull/15937/commits/a40cc0d18b07e410125517e3abb1116257b77433 . I.e., with my patch, POSIX-using apps no longer need to do `target_include_directories(app PRIVATE $ENV{ZEPHYR_BASE}/include/)` (which is kinda implementation detail of POSIX subsys, which should not leak into config of each and every user app).

As can be seen, CI is green for this PR, so hopefully teh handling above is not broken. Well, I'd still hope to try it manually myself to be double-sure (tomorrow, perhaps).

But otherwise, I'd leave it to @galak, the author of  https://github.com/zephyrproject-rtos/zephyr/pull/14312, to vote this PR up/down.

**Comment by pabigot** [comment]
https://reviews.llvm.org/D32411 includes a commit that documents a design to support `#include_next` on MSVC, which doesn't have it, using nasty looking code and CMake.  However, the practice within [newlib itself](https://sourceware.org/git/gitweb.cgi?p=newlib-cygwin.git;a=blob;f=newlib/libc/include/limits.h;h=893f10834f6c13947c71785371a7d55c84115937;hb=refs/heads/master#l19) is to simply provide the missing content inline.

Are there non-clang/gnuc toolchains we need to worry about?

File: CODEOWNERS (hunk 2 above)
Comment (pabigot) [reply CODEOWNERS]: I've never liked using `cpp` for C++; I usually use `cxx` but `cpp` was existing practice.  I've renamed it to `libcxx`.

File: lib/libc/newlib/include/stdint.h (hunk 1 above)
Comment (pabigot) [reply lib/libc/newlib/include/stdint.h]: I found it in [clang](http://clang.llvm.org/docs/LanguageExtensions.html#has-include-next) when checking whether that supported `#include_next`.  I wasn't aware that it was also supported in GNUC.

After further review I've decided the protection stuff was unnecessarily complex, and replaced it with a bare use of `#include_next` and a comment saying what needed to be done if some toolchain doesn't support that capability.

File: CODEOWNERS (hunk 2 above)
Comment (pfalcon) [reply CODEOWNERS]: > I've renamed it to libcxx.

That's definitely better, thanks.

**Comment by pfalcon** [comment]
Ok, manually tested tests/posix/common, tests/posix/fs - all's ok (yeah, that should be covered by CI, but as we know there's a bit randomization and subsetting in it, so some tests aren't run all platforms, and given a trouble it took to lead POSIX-related changes into the mainline, I wanted to have an explicit checkpoint here).

So, I'd be happy to +1 this, just giving some time to @galak to review too.

**Review by SebastianBoe — APPROVED** [review]
Great work.

**Review by pfalcon — APPROVED** [review]

**Review by galak — APPROVED** [review]

**Pull request merged**